### PR TITLE
Speed up loading of MOSViz data

### PR DESCRIPTION
**Pull request by astrofrog**
In glue, by default the link graph is re-computed every time a dataset is added. We can speed up the loading of MOSViz data significantly by waiting until the end to update the link graph. This requires a change in glue-core:

* https://github.com/glue-viz/glue/pull/2225
- [x] Bump glue-core minversion after the PR above is merged and released upstream.

Might also need this:

* astrofrog/jdaviz#4

Code used to profile the loading:

```python
from jdaviz.configs.mosviz.helper import MosViz

mosviz = MosViz()
mosviz.app
mosviz.load_niriss_data('NIRISS_for_parser_p0171')
```

Before:

```
real	1m57.610s
user	1m58.607s
sys	0m1.159s
```

After:

```
real	0m6.107s
user	0m7.259s
sys	0m1.074s
```

However, there is still scope for optimization. It looks like ``Application._link_new_data`` is actually wrong in a number of cases - 
it simply links the first world coordinate of datasets which results in e.g. the spectral axis of a 1-d spectrum being linked with the declination axis of cutouts, so there are too many links in the link graph, some of them wrong.

We should therefore figure out what we actually want ``Application._link_new_data`` to do in the general case. At the moment this method seems dangerous and it would be safer to set up links in the individual parsers, or find a more general/correct way to do this. I wonder if it was added back in the early days when specviz was the main application - in which case it would have been correct (except in cases where one spectrum is in wavelength and one is in frequency).

### CODE REVIEW

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/parsers.py:429]: Why not just use `dict`? It is ordered now, right?

**Comment by pllim** [comment]
p.s. Tsk tsk, you removed the PR checklist? 😉

**Comment by astrofrog** [comment]
Oops didn't even notice the checklist when I opened the PR, sorry!

File: jdaviz/configs/mosviz/plugins/parsers.py
Comment (pllim) [reply jdaviz/configs/mosviz/plugins/parsers.py:429]: Yes.

https://github.com/spacetelescope/jdaviz/blob/8019780dcb4d9198b2bb4b1b5a0d7392d4508969/setup.cfg#L18

Comment (astrofrog) [reply jdaviz/configs/mosviz/plugins/parsers.py]: Done

**Comment by astrofrog** [comment]
I think https://github.com/spacetelescope/jdaviz/pull/762 might do the right thing in terms of fixing the links themselves (and doing them at the end makes sense anyway)

**Comment by pllim** [comment]
Now I am confused... Should #762 build on this PR, or _replaces_ this PR?

**Comment by astrofrog** [comment]
We need to combine them. Maybe I can do a PR to

**Comment by astrofrog** [comment]
Alternatively we can merge one and rebase the other, the order is not important

**Comment by astrofrog** [comment]
I have an idea, more soon

**Comment by astrofrog** [comment]
Ok so I've temporarily included the commits from #762 in here to make it easier to test things together and I end up with a load time for the NIRISS data of 1.65 seconds:

Screenshot from 2021-08-12 20-41-32

However this PR doesn't change anything about the correctness of #762.

There are two main paths at this point:

* We merge this current PR and then rebase #762 and then add a commit like https://github.com/spacetelescope/jdaviz/pull/782/commits/f6b8c2c5cfa61b094a445f5ba1429092652ede67

* We focus on getting #762 correct even if slower, and then I can rebase this PR here to make it faster

Doing the first option would make it easier to then work on #762 since then things would be faster so easier to try things out. If people are on board with that I can do a release of glue-core in the next 2-3 days and we could then merge this after I update the glue-core pinning.

@ojustino @pllim @rosteen - what do you think?

**Comment by pllim** [comment]
I like option 1, but I think @ojustino should have the final say, given that he took over this work from Jesse.

**Comment by ojustino** [comment]
I was hoping to review more extensively before commenting, but I'll say I'm OK with merging this and rebasing #762. I'm working through some install issues and will test once they're resolved.

**Comment by astrofrog** [comment]
I'll remove the commits from #762 now to make it easier to review - it will take a few days for the glue-core release to happen in all likelihood so there will still be time to review this.

**Comment by astrofrog** [comment]
This is now ready for review and can be merged if approved as glue-core 1.2 is out.

**Comment by astrofrog** [comment]
This actually just speeds up the ``mos_niriss_parser`` - I need to apply a similar change to the other parsers, I will try and do that very shortly.

**Comment by pllim** [comment]
👍 

Out of scope but the codecov complains concern me. This parser should be tested more thoroughly...

**Comment by astrofrog** [comment]
Agree that we should really add tests for the parsers!

**Comment by rosteen** [comment]
Catching up after a short vacation, I also think that option 1 (review and merge this, then rebase #762 ) is the way to go. I'll try to review this between meetings today.

**Review by ojustino — APPROVED** [review]
I just got this working locally. Since the delay link manager was merged upstream, I tested with the latest version of glue on `master`.

I get 8.5 seconds for `load_niriss_data()` in `notebooks/MosvizNIRISSExample.ipynb` and 13 seconds for 20 spectra + images with `load_data()` -- much better. Problems mentioned in #762 with subset selection still persist, but those are for solving elsewhere. I don't see other glaring problems.

**Comment by rosteen** [comment]
I just submitted a PR to @astrofrog 's fork to add a NIRISS parser test, that might make Codecov happy.

**Review by rosteen — APPROVED** [review]
Test passed, Codecov is happy. Approved, will merge in a moment.

**Pull request merged**
